src/utils/tcti.rs

```diff
@@ -0,0 +1,451 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+use crate::response_code::{Error, WrapperErrorKind};
+use regex::Regex;
+use std::convert::TryFrom;
+use std::ffi::CString;
+use std::net::IpAddr;
+use std::path::PathBuf;
+use std::str::FromStr;
+
+const DEVICE: &str = "device";
+const MSSIM: &str = "mssim";
+const TABRMD: &str = "tabrmd";
+
+// Possible TCTI to use with the ESYS API.
+/// Placeholder TCTI types that can be used when initialising a `Context` to determine which
+/// interface will be used to communicate with the TPM.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Tcti {
+    /// Connect to a TPM available as a device node on the system
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Device_Init)
+    Device(DeviceConfig),
+    /// Connect to a TPM (simulator) available as a network device
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Mssim_Init)
+    Mssim(MssimConfig),
+    /// Connect to a TPM through an Access Broker/Resource Manager daemon
+    ///
+    /// For more information about configuration, see [this page](https://www.mankier.com/3/Tss2_Tcti_Tabrmd_Init)
+    Tabrmd(TabrmdConfig),
+}
+
+impl TryFrom<Tcti> for CString {
+    type Error = Error;
+
+    fn try_from(tcti: Tcti) -> Result<Self, Error> {
+        let tcti_name = match tcti {
+            Tcti::Device(..) => DEVICE,
+            Tcti::Mssim(..) => MSSIM,
+            Tcti::Tabrmd(..) => TABRMD,
+        };
+
+        let tcti_conf = match tcti {
+            Tcti::Mssim(config) => {
+                if let Some(ServerAddress::Hostname(name)) = &config.host {
+                    if !hostname_validator::is_valid(name) {
+                        return Err(Error::WrapperError(WrapperErrorKind::InvalidParam));
+                    }
+                }
+                format!(
+                    ":host={},port={}",
+                    config.host.unwrap_or_default(),
+                    config.port.unwrap_or(2321)
+                )
+            }
+            Tcti::Device(DeviceConfig { path: Some(path) }) => format!(
+                ":{}",
+                path.to_str()
+                    .ok_or(Error::WrapperError(WrapperErrorKind::InvalidParam))?,
+            ),
+            Tcti::Device(DeviceConfig { path: None }) => String::default(),
+            Tcti::Tabrmd(config) => format!(
+                ":bus_name={},bus_type={}",
+                config
+                    .bus_name
+                    .unwrap_or_else(|| String::from("com.intel.tss2.Tabrmd")),
+                config.bus_type.unwrap_or(BusType::System)
+            ),
+        };
+
+        CString::new(format!("{}{}", tcti_name, tcti_conf))
```

Comment:
Sorry for been 'that guy' here but. I think it would be clearer if you left out the limiter when you create tcti_conf and add it here in the format instead.
```rust
CString::new(format!("{}:{}", tcti_name, tcti_conf))
```

Reply:
The problem with that is when `tcti_conf` is empty. I could branch before, I guess

Reply:
Done